state-synchronizer/src/request_manager.rs

```diff
@@ -234,12 +234,47 @@ impl RequestManager {
         None
     }
 
-    pub fn pick_peers(&self) -> Vec<PeerNetworkId> {
-        self.eligible_peers
-            .iter()
-            .take(self.multicast_level)
+    pub fn pick_peers(&mut self) -> Vec<PeerNetworkId> {
+        // Strategy: pick peers using multicast level
+        // if no live peers exist for this multicast level, keep failing over to next level
+
+        // find peers in all networks for the current multicast level
+        let peers = self
+            .eligible_peers
+            .range((Bound::Unbounded, Included(self.multicast_level)))
             .filter_map(|(_, (peers, weighted_index))| Self::pick_peer(peers, weighted_index))
-            .collect::<Vec<_>>()
+            .collect::<Vec<_>>();
+
+        if !peers.is_empty() {
+            return peers;
+        }
+
+        // if no peers were found from current multicast level above,
+        // do best effort search for next network with live peers to fail over to and pick a peer from it
+        // and update multicast level accordingly
+        let failover = self
+            .eligible_peers
+            .range((Excluded(self.multicast_level), Bound::Unbounded))
+            .next()
```

Comment:
this doesn't seem doing what the comment suggests? is the `.next()` guaranteed to have live peers or you actually want to use `.find_map()`?

Reply:
`eligible_peers` is always going to contain live peers - see `update_peer_selection_data`

Reply:
I think this fn can be simpler
```
  while peers.is_empty() || level < self.multicast_level {
    let (level, p) = eligible_peers_iter.next().map(..);
    peers.extend(&mut p);
  }
  if level > self.multicast_level {
    ...
  }
  peers
```

Reply:
pushed new commit with a variation of the above (reorganize the nesting levels to avoid iter/borrow craziness)